BuiltIn Metrics

bradbeam opened the pull request:
Wanted to get stuff out there so we could use this for discussion. 

Currently I'm thinking of introducing a new metric type ( MetricBuiltIn ) to facilitate our ability to include metrics out of the box.

This will eventually expand to:
- include a prometheus deployment
- include built in metrics
  - these are ultimately prometheus metrics, but allow for generating the prometheus server url and inject additional metadata ( label selectors, query ranges, etc )
- incorporate the built in metrics automatically as part of a trial run

The workflow would be something like:
- experiment gets created
- prometheus server and kube state metrics get deployed alongside experiment
- during trial run, prometheus server scrapes cadvisor and kube state metrics
- after trial run, user provided metrics and builtin metrics are evaluated
- profit(?)

bradbeam committed:
feat: Introduce BuiltIn metric type

This includes a new builtin metric type so we can begin to include our own
queries.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add support for identifying local prometheus url

This adds a new template function to generate the prometheus url based on the trial name.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add stubs for builtin metric types

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add utilization query and handle builtin metric

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
fix: Fixes from testing

- Update query to use 1s step time for calculating total limit
- Fix where we inject builtin metrics so the experiment in the controller and server align properly
- some logging that should be removed

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
update cpu util calculation to use sum to account for replicas

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
fix up template test

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add annotation to make label query dynamic

This introduces the redskyops.dev/metric-target annotation to supply the label=value information
we need when running the builtin utilization queries.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
chore: Removing quickstart until we figure out the prometheus situation

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

jgustie commented on internal/template/template.go:
```diff
@@ -52,6 +53,8 @@ type MetricData struct {
 	// List of pods from the trial namespace (only available for "pods" type metrics)
 	Pods *corev1.PodList
+	// The experiment identifier
+	ExperimentNamespacedName types.NamespacedName
```
Can we just call this `Experiment`? You can reference it in templates as `.Experiment.Name`, `.Experiment.Namespace` or `.Experiment` (which is the "namespace/name" form).

bradbeam replied:
This has been dropped. 
With the change to a static name for the service, we can use `.Trial` to get the namespace

jgustie commented on internal/metric/builtin.go:
```diff
@@ -0,0 +1,54 @@
+    on (pod) group_left sum_over_time(kube_pod_container_resource_limits_cpu_cores[1h:1s])
+  )
+)
+`
```
Why isn't this just a template function or a named template that you can use with the existing Prometheus collector? For example, `query: '{{ cpuUtilization "component=redis,component=web" }}'` would effectively evaluate out to the CPU utilization query for pods labeled with either `component=redis` or `component=web`.

jgustie commented on internal/template/template_test.go:
```diff
@@ -156,14 +215,19 @@ func TestEngine(t *testing.T) {
+				got, err = eng.RenderHelmValue(inputType, tc.trial)
 			case *redskyv1beta1.Metric:
-				got, _, err = eng.RenderMetricQueries(tc.input.(*redskyv1beta1.Metric), tc.trial, tc.obj)
+				// Not sure of a better way to handle this
```
I think we should probably start by splitting `TestEngine` into `TestEngine_RenderHelmValue`, `TestEngine_RenderMetricQueries`, and `TestEngine_RenderPatch`.

jgustie commented on internal/template/template.go:
```diff
@@ -144,6 +149,25 @@ func (e *Engine) RenderMetricQueries(metric *redskyv1beta1.Metric, trial *redsky
+
+	return labelPairBytes.String(), err
+}
+
```
How are these getting used? I don't think it necessary to leverage templates for this.

With the existing Prometheus metric collection, just change it so if there is nothing specified (i.e. today no labels, in the future, no URL) we just default the Prometheus URL to what we would be standing up. That is, instead of failing when we see missing configuration we will push the potential failure further down the road (when we try to connect).

jgustie reviewed (changes requested):
Looks like there are also comments, in progress printlns etc. that need some clean up.

bradbeam committed:
refactor: change generated prometheus url to be well known

This changes the generated prometheus url from `prom-experimentname` to `rso-prometheus`
to simplify the identification of the prometheus server.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add support for deleting timeseries

This change introduces the ability to delete a set of timeseries from prometheus.
This allows us to reset the data to increase the accuracy and simplify the queries we will run.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
chore: cleanup unnecessary log statements and comments

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
refactor: Remove promServer template function

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add prometheus deployment to setuptools image

This introduces another argument for the setuptools container to handle
a prometheus installation.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add experiment setup job

This adds in the ability for the server controller to initiate a setup job
as part of the experiment lifecycle.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add support for deleting experiment job

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
fix: Change setuptools create to use apply

This allows us to handle resources that may already exist.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add additional service account + rbac for builtins

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
chore: Drop unnecessary permissions for kube-state-metrics

Since we only need to grab the labels on the pod, we can drop all the rest of this noise
so we can reduce the permissions necessary on the builtin service account.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
fix: Update prometheus url to remove namespace

Since we are deploying prometheus stuff to redsky-system, we should take out
the reference to the trial namespace.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Reenable namespace based prometheus deployment

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Update cpu utilization query

This removes the filter for running pods, so we should be able to handle HPA trials.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>
